CredentialsProvider API should be asynchronous

tsegismont opened the pull request:
The synchronous method is deprecated and is invoked by the asynchronous variant.

Extensions must be updated to invoke the asynchronous variant.

- Closes: #32021

tsegismont commented:
> I like the change (and actually need it for the mailer), but I think a few usages have been overlooked.

Thanks for the feedback, I'll looking into the failures now

tsegismont commented on extensions/credentials/runtime/src/main/java/io/quarkus/credentials/CredentialsProvider.java:
```diff
@@ -12,11 +19,25 @@ public interface CredentialsProvider {
+     */
+    default Uni<Map<String, String>> getCredentialsAsync(String credentialsProviderName) {
+        return Uni.createFrom().item(() -> getCredentials(credentialsProviderName))
+                .runSubscriptionOn(Infrastructure.getDefaultExecutor());
```
The problem was not with the default implementation of `getCredentialsAsync`, but with the synchronous nature of `io.quarkus.oidc.common.runtime.OidcCommonUtils#fromCredentialsProvider` even though this method is invoked on an event loop thread.

I reverted the change in this method and added a comment. Ideally, the problem would be addressed in a follow-up PR, but I'm not knowledgeable with this area. I could help with a PR review though

sberyozkin replied:
Thanks @tsegismont.
When I get to that, I'll definitely cc you.
IMHO though, it would still be better to avoid setting a particular thread pool in the default implementation,  since its role is to avoid breaking existing code, not to offer an actual async implementation. 
Custom extensions or code that has no async CP requirements or immediate capabilites may have no other option but to keep the deprecated method ref.
IMHO your original update to OidcCommonUtils should've worked as you typed it.

In any case, thanks for the PR

sberyozkin reviewed (approved):
Thanks @tsegismont

sberyozkin commented on extensions/credentials/runtime/src/main/java/io/quarkus/credentials/CredentialsProvider.java:
```diff
@@ -12,11 +19,25 @@ public interface CredentialsProvider {
+     * @deprecated invoke {@link #getCredentialsAsync(String)} instead
      */
-    Map<String, String> getCredentials(String credentialsProviderName);
+    @Deprecated
+    default Map<String, String> getCredentials(String credentialsProviderName) {
```
@tsegismont Hi,

>Having a single async method allows to solve both problems, but it is true that some people prefer to stay away from reactive. A separate interface (BlockingCredentialsProvider ?) could do the trick.

It is certainly not me :-), I'll have no problems updating the OIDC code (not immediately, but I opened an issue with CC to you), I think your new interface method is perfect, my only reservation is that if someone has a blocking call working for whatever reason, then, ideally, the same blocking call expressed in Uni `await().indefinitely()` terms should also work to allow a simple move away from the deprecated method. 

What if the the consumer of the API already running on a ManagedExecutor worker thread, in a blocking mode, where a custom CP can have some request scoped bean injected... 

In any case, apologies, I'm not really an expert in how to link various threads together across all the chain call, so please ignore me if I talk some nonsense :-)

I'm going quiet now, as I'm in Oot of office mode 
Thanks

geoand commented:
@yrodiere any last comments?

tsegismont commented:
> So, no I discovered the PR. As I said as comment, I would not deprecate the blocking API, but implement one on top of the other (using a default method).

The implementation does exactly this, so we're good on that front.

No problem, I'll remove the deprecation part

tsegismont committed:
CredentialsProvider API should be asynchronous

Closes #32021

By default, the synchronous method is invoked by the asynchronous variant on a worker thread.

Extensions must be updated to invoke the asynchronous variant.

cescoffier reviewed (commented):
For me the question is how the credential provider is intended to be implemented and used.

For the usage, generalizing async and _awaiting_ if needed is accepetable.
However, to do that, the provider needs to be implemented using an async API. 

I would not deprecate the blocking API, but instead try to find a way to have both, with one using the other if needed. We cool imagine adding a new interface for async credential providers if needed. 

For blocking (or synchronous) async provider, extensions should be careful and call them from a different thread if they are on the event loop.

tsegismont commented:
@cescoffier in the last commit, I removed the deprecation and the synchronous method is invoked from the async one (as discussed in https://github.com/quarkusio/quarkus/issues/32021#issuecomment-2493069814)

So, if I understand correctly, you changed your mind and would prefer separate interfaces for async/sync credential providers?

It's fine with me, I just want to make sure we agree on the desired implementation before changing the PR.

cescoffier commented:
Can we have both in the same interface?

michalvavrik reviewed (commented):
I think this PR will have unfortunate impact on performance, because all existing credentials providers produced by users are going to be executed on the worker thread. Especially in security area, our code is most often executed on the IO thread thanks to how early we do stuff.

What I am missing in this PR:

- migration guide doesn't mention this change https://github.com/quarkusio/quarkus/wiki/Migration-Guide-3.30
- Quarkus documentation doesn't provide clue which method should be used, users should understand that they only need to use `getCredentials` (that is the synchronous method) when they need to block
- javadoc should remind users that the `getCredentials` is only for blocking executions, otherwise they are going to pay a price

I just think that assuming that every user credentials provider created until now is blocking cannot be right.
